### DISCUSSION TITLE
Routing weirdness when nesting

### DISCUSSION THREAD

**Discussion by szmarci** [remix-run/remix#1825] [Q&A] [answered]
I came across a weird thing I don't really understand regarding routing. 
I made a small repo in case you want to reproduce it: [https://github.com/szmarci/remix-routing](https://github.com/szmarci/remix-routing)

I have 3 components, Blue, Yellow and Red.
I have 2 subroutes `/colors` `/colors/yellowandred`.

Color components are really simple, just a color name, some style and class:
```
export const Yellow = () => {
  return (
    <div className="yellow" style={{backgroundColor: "yellow"}}>
      yellow
    </div>
  )
}
```

When we hit `/colors` I render a button and the Blue component:
<img width="534" alt="Screenshot 2022-02-06 at 23 39 45" src="https://user-images.githubusercontent.com/1974290/152704555-bc4cf8a4-24b2-433b-aee5-d4ffb439c9f8.png">

So far so good. The button will route to `/colors/yellowandred` and I also pass in some state to the route.
`navigate("yellowandred", {state: {colorid: "red"}});`
I read this value from the url when we arrive at `/colors/yellowandred` and conditionally render red or blue:
<img width="247" alt="Screenshot 2022-02-06 at 23 44 03" src="https://user-images.githubusercontent.com/1974290/152704721-56292041-9217-4383-b1be-4fc9b456ffac.png">

The result:

![image](https://user-images.githubusercontent.com/1974290/152704626-9a09042f-c4ad-46c1-9ecb-d52dd23585d0.png)

Still okay, yellow and red are rendered as expected. 
Now when I refresh it gets weird:
![image](https://user-images.githubusercontent.com/1974290/152704881-53ef2a45-ad11-498a-b2d7-505f4016c44f.png)

At first it flashes with the text `blue` for a moment, then it immediately swaps it to `red`, but it is keeping `blue`'s class and style.
How is this possible?

**Comment by mkrtchian** [comment, marked as answer]
This seems to be a hydration problem (mismatch between the React code rendered in the server, and the one rendered in the client).
When reloading with the weird behavior I get this error:
> Warning: Prop `className` did not match. Server: "blue" Client: "red"

The **Location state** you are using is explained in the [React Router V6 documentation here](https://reactrouter.com/docs/en/v6/getting-started/concepts#locations).
It is created on client side, and seems to be kept by your browser until you change the URL, or put the same URL by hand after changing it.

I guess that as Remix does not know that information on server side. When you hit reload, the server renders your page first thinking it's the Blue component. Then that comes to the client and is displayed immediately by the browser as Blue. And when the client side JavaScript loads it "hydrates" the existing DOM with the client side React code, which results in something a bit different: the Red component. And only a part of it is changed, the `className` is kept with a warning.

I don't know if this has to be considered as a bug or not :thinking: Maybe Remix could send the `state` information to the server when it is set ?

**Reply by ccssmnn** [reply]
Exactly this.

When you want the setting to be preserved on a refresh, consider storing the state in a session (cookie or database). Then, each loader and action can read the cookie on the server and return the result.

**Reply by szmarci** [reply]
Well, thanks. I don't currently have any loader, action, session, whatnot in this app. But in essence I have to store the state info in a cookie when pushing the button on `/colors` and when going to `/colors/yellowandred` that cookie must be sent along with the request to the loader and from the loader I have to return that state to the component and use it in the conditional rendering, right?

Btw, I noticed I could hack it by changing the Red component markup from `div` to `span`, so I guess hydration cannot "merge" them when they are of different type. I mean it still flashes for a moment with the server side version, but then correctly changes to the client side version.

**Reply by ccssmnn** [reply]
Yes. See [the sessions chapter in the docs](https://remix.run/docs/en/v1/api/remix#sessions) for guidance.
